---
title: "KAOs WIP"
author: "Alexander Engell-Hansen"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Loading packages:
```{r}
# library(ggplot)
library(GenomicRanges)
library(rtracklayer)
library(usethis) 
usethis::edit_r_environ() # R_MAX_VSIZE=100Gb
```



Initialising the data:
```{r}
# To-do: Subset on import 
setwd("/Users/alexanderengell-hansen/Desktop/git/dataprojekt/data")
annot_gr = import('annotation_subset.gtf') # Annotations
bw = import('control.bw') # Control data
bw_cps = import('cps.bw') # Test data
```


Data transformation:
```{r}
# Transciption on GAPDH genome
annot_gr <- subset(annot_gr, gene_name == 'GAPDH')
ts_annot_gr <- subset(annot_gr, type == "transcript")

# Chromosome 12
bw_12 <- subset(bw, seqnames == "chr12")
bw_cps_12 <- subset(bw_cps, seqnames == "chr12")
```




Create DF with transcript no. column, position (start & end), score etc.:
```{r}
subsetOverlaps_test <- as.list(rep(0,11))
subsetOverlaps_control <- as.list(rep(0,11))

for (i in seq_along(ts_annot_gr)) {
	current_observation <- ts_annot_gr[i]
  
	start_coords <- start(current_observation) - 500
	end_coords <- end(current_observation) + 100000
  
	new_ranges <- GRanges(
    	seqnames = "chr12",
    	ranges = IRanges(start = start_coords, end = end_coords)
  	)
  
	subsetOverlaps_test[i] <- (subsetByOverlaps(bw_cps_12, new_ranges))
	subsetOverlaps_control[i] <- (subsetByOverlaps(bw_12, new_ranges))
}
```

Create an empty data frame to store the results:
```{r}
result_df_test <- data.frame(transcript_number = numeric(),
                        seqnames = character(),
                        start = numeric(),
                        end = numeric(),
                        score = numeric(),
                        stringsAsFactors = FALSE)

result_df_control <- data.frame(transcript_number = numeric(),
                        seqnames = character(),
                        start = numeric(),
                        end = numeric(),
                        score = numeric(),
                        stringsAsFactors = FALSE)


overlap_values_list_test <- subsetOverlaps_test
overlap_values_list_control <- subsetOverlaps_control
```

Loop over each GRanges object in the list:
```{r}
for (i in seq_along(overlap_values_list_test)) {
	# Extract the current GRanges object
	current_gr <- overlap_values_list_test[[i]]
  
	# Extract range and score values from the current GRanges object
	range_values <- as.data.frame(current_gr)
  
	# Add transcript number column
	range_values$transcript_number <- i

	# Append range and score values to the result data frame
	result_df_test <- rbind(result_df_test, range_values)
}

result_df_test$transcript_number <- as.factor(result_df_test$transcript_number)

for (i in seq_along(overlap_values_list_control)) {
	# Extract the current GRanges object
	current_gr <- overlap_values_list_control[[i]]
  
	# Extract range and score values from the current GRanges object
	range_values <- as.data.frame(current_gr)
  
	# Add transcript number column
	range_values$transcript_number <- i

	# Append range and score values to the result data frame
	result_df_control <- rbind(result_df_control, range_values)
}

result_df_control$transcript_number <- as.factor(result_df_control$transcript_number)
```


FILL WITH 0's:
```{r}
fill_gaps <- function(df) {
    # Initialize an empty list to store the new rows
    new_rows <- list()
    for(i in 1:(nrow(df) - 1)) {
        current_position <- df$start[i]
        next_position <- df$start[i + 1]
        if(next_position - current_position > 1) {
            # Generate the missing positions
            missing_positions <- (current_position + 1):(next_position - 1)
            for(missing_pos in missing_positions) {
                new_row <- df[i, ]
                new_row$start <- missing_pos
                new_row$score <- 0 # Set score to 0 for the new row
                new_rows[[length(new_rows) + 1]] <- new_row
            }
        }
    }   

    if(length(new_rows) > 0) {
        new_rows_df <- do.call(rbind, new_rows)
        # Combine with the original dataframe and sort
        df <- rbind(df, new_rows_df)
        df <- df[order(df$start), ]
    }

    return(df)
}
```

FILL WITH 0's USAGE:
```{r}
# USAGE
df_filled_test <- fill_gaps(result_df_test)
print(head(df_filled))
```


```{r}
fill_gaps <- function(df) {
  # Initialize an empty list to store the new rows
  new_rows <- list()
  for(i in 1:(nrow(df) - 1)) {
      current_position <- df$start[i]
      next_position <- df$start[i + 1]
      if(next_position - current_position > 1) {
        # Generate the missing positions
        missing_positions <- (current_position + 1):(next_position - 1)
        for(missing_pos in missing_positions) {
          new_row <- df[i, ]
          new_row$start <- missing_pos
          new_row$score <- 0 # Set score to 0 for the new row
          new_rows[[length(new_rows) + 1]] <- new_row
        }
      }
    }
    if(length(new_rows) > 0) {
        new_rows_df <- do.call(rbind, new_rows)
        # Combine with the original dataframe and sort
        df <- rbind(df, new_rows_df)
        df <- df[order(df$start), ]
      }
  
    return(df)
}
```



AGGREGATE DF WITH BINSIZE:
```{r}
# AGGREGATE DF WITH BINSIZE

create_aggregated_df <- function(df, bin_size) {
	# Determine the number of chunks
	n <- nrow(df)
	num_chunks <- ceiling(n / bin_size)



	medians <- numeric(num_chunks)
	means <- numeric(num_chunks)

	for(i in 1:num_chunks) {
		# Calculate indices for slicing the dataframe
		start_index <- (i - 1) * 100 + 1
		end_index <- min(i * 100, n)

		chunk <- df[start_index:end_index, ]
		medians[i] <- median(chunk$position)
		means[i] <- mean(chunk$score)
		}	


	aggregated_df <- data.frame(start = medians, score = means)

	return(aggregated_df)
}

# USAGE

# new_df <- create_aggregated_df(df_filled)
# print(new_df)
```




```{r}

# Get diffs:
diffs <- c(0, diff(new_df$score))
new_df$diff <- diffs 

# Getting start and termination position:
krop_start <- subset(new_df, diff == max(diffs))$start - 100
krop_slut <- subset(new_df, diff == min(diffs))$start
```

